tls/s2n_config.c

```diff
@@ -135,6 +139,53 @@ static int s2n_config_cleanup(struct s2n_config *config)
     GUARD(s2n_config_free_dhparams(config));
     GUARD(s2n_free(&config->application_protocols));
     GUARD(s2n_array_free(config->cert_and_key_pairs));
+    GUARD(s2n_map_free(config->cert_and_key_pair_map));
+
+    return 0;
+}
+
+static int s2n_config_update_cert_and_key_pair_map(struct s2n_map *cert_and_key_pair_map,
+        struct s2n_blob *name,
+        struct s2n_cert_chain_and_key *cert_key_pair)
+{
+    /* s2n_map does not allow zero-size key */
+    if (name->size == 0) {
+        return 0;
+    }
+    struct s2n_blob map_value = { 0 };
+    s2n_authentication_method auth_method = s2n_cert_chain_and_key_get_auth_method(cert_key_pair);
+    if (s2n_map_lookup(cert_and_key_pair_map, name, &map_value) == 0) {
+        struct s2n_cert_chain_and_key_map_value value = {{ 0 }};
+        value.certs[auth_method] = cert_key_pair;
+        map_value.data = (uint8_t *) &value;
+        map_value.size = sizeof(struct s2n_cert_chain_and_key_map_value);
+
+        GUARD(s2n_map_unlock(cert_and_key_pair_map));
+        GUARD(s2n_map_add(cert_and_key_pair_map, name, &map_value));
+        GUARD(s2n_map_complete(cert_and_key_pair_map));
+    } else {
+        struct s2n_cert_chain_and_key_map_value *value = (void *) map_value.data;;
+        if (value->certs[auth_method] == NULL) {
```

Comment:
Yes, it is for the case where certs with different key types but with the same domain name are loaded. I added integration test cases for it.

For now we just skip adding the cert to the map and return 0 if `value->certs[auth_method] != NULL` since it is not failure/error. If we handle this case differently in the future, we can return other values accordingly.